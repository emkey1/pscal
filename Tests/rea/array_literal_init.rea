int main() {
  int numbers[3] = [1, 2, 3];
  str labels[2] = ["alpha", "beta"];
  writeln(numbers[1]);
  writeln(labels[0]);
  return 0;
}

main();
