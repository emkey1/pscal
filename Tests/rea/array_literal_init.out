{

  "node_type": "PROGRAM",
  "var_type_annotated": "VOID",
  "main_block": 
  {

    "node_type": "BLOCK",
    "var_type_annotated": "VOID",
    "is_global_scope": false,
    "declarations": 
    {

      "node_type": "COMPOUND",
      "var_type_annotated": "VOID",
      "children": [
        {

          "node_type": "FUNCTION_DECL",
          "token": {
              "type": "IDENTIFIER",
              "value": "main"
          },
          "var_type_annotated": "INT64",
          "is_inline": false,
          "right": 
          {

            "node_type": "TYPE_IDENTIFIER",
            "token": {
                "type": "IDENTIFIER",
                "value": "int"
            },
            "var_type_annotated": "INT64"
          }
,
          "extra": 
          {

            "node_type": "COMPOUND",
            "var_type_annotated": "VOID",
            "children": [
              {

                "node_type": "VAR_DECL",
                "var_type_annotated": "ARRAY",
                "left": 
                {

                  "node_type": "ARRAY_LITERAL",
                  "var_type_annotated": "ARRAY",
                  "children": [
                    {

                      "node_type": "NUMBER",
                      "token": {
                          "type": "INTEGER_CONST",
                          "value": "1"
                      },
                      "var_type_annotated": "INT64",
                      "i_val": 0
                    },
                    {

                      "node_type": "NUMBER",
                      "token": {
                          "type": "INTEGER_CONST",
                          "value": "2"
                      },
                      "var_type_annotated": "INT64",
                      "i_val": 0
                    },
                    {

                      "node_type": "NUMBER",
                      "token": {
                          "type": "INTEGER_CONST",
                          "value": "3"
                      },
                      "var_type_annotated": "INT64",
                      "i_val": 0
                    }
                  ]
                }
,
                "right": 
                {

                  "node_type": "ARRAY_TYPE",
                  "var_type_annotated": "ARRAY",
                  "right": 
                  {

                    "node_type": "TYPE_IDENTIFIER",
                    "token": {
                        "type": "IDENTIFIER",
                        "value": "int"
                    },
                    "var_type_annotated": "INT64"
                  }
,
                  "children": [
                    {

                      "node_type": "SUBRANGE",
                      "var_type_annotated": "INT64",
                      "left": 
                      {

                        "node_type": "NUMBER",
                        "token": {
                            "type": "INTEGER_CONST",
                            "value": "0"
                        },
                        "var_type_annotated": "INT64",
                        "i_val": 0
                      }
,
                      "right": 
                      {

                        "node_type": "NUMBER",
                        "token": {
                            "type": "INTEGER_CONST",
                            "value": "2"
                        },
                        "var_type_annotated": "INT64",
                        "i_val": 2
                      }

                    }
                  ]
                }
,
                "children": [
                  {

                    "node_type": "VARIABLE",
                    "token": {
                        "type": "IDENTIFIER",
                        "value": "numbers"
                    },
                    "var_type_annotated": "ARRAY"
                  }
                ]
              },
              {

                "node_type": "VAR_DECL",
                "var_type_annotated": "ARRAY",
                "left": 
                {

                  "node_type": "ARRAY_LITERAL",
                  "var_type_annotated": "ARRAY",
                  "children": [
                    {

                      "node_type": "STRING",
                      "token": {
                          "type": "STRING_CONST",
                          "value": "alpha"
                      },
                      "var_type_annotated": "STRING"
                    },
                    {

                      "node_type": "STRING",
                      "token": {
                          "type": "STRING_CONST",
                          "value": "beta"
                      },
                      "var_type_annotated": "STRING"
                    }
                  ]
                }
,
                "right": 
                {

                  "node_type": "ARRAY_TYPE",
                  "var_type_annotated": "ARRAY",
                  "right": 
                  {

                    "node_type": "TYPE_IDENTIFIER",
                    "token": {
                        "type": "IDENTIFIER",
                        "value": "str"
                    },
                    "var_type_annotated": "STRING"
                  }
,
                  "children": [
                    {

                      "node_type": "SUBRANGE",
                      "var_type_annotated": "INT64",
                      "left": 
                      {

                        "node_type": "NUMBER",
                        "token": {
                            "type": "INTEGER_CONST",
                            "value": "0"
                        },
                        "var_type_annotated": "INT64",
                        "i_val": 0
                      }
,
                      "right": 
                      {

                        "node_type": "NUMBER",
                        "token": {
                            "type": "INTEGER_CONST",
                            "value": "1"
                        },
                        "var_type_annotated": "INT64",
                        "i_val": 1
                      }

                    }
                  ]
                }
,
                "children": [
                  {

                    "node_type": "VARIABLE",
                    "token": {
                        "type": "IDENTIFIER",
                        "value": "labels"
                    },
                    "var_type_annotated": "ARRAY"
                  }
                ]
              },
              {

                "node_type": "EXPR_STMT",
                "var_type_annotated": "VOID",
                "left": 
                {

                  "node_type": "WRITELN",
                  "var_type_annotated": "UNKNOWN_VAR_TYPE",
                  "children": [
                    {

                      "node_type": "ARRAY_ACCESS",
                      "var_type_annotated": "INT64",
                      "left": 
                      {

                        "node_type": "VARIABLE",
                        "token": {
                            "type": "IDENTIFIER",
                            "value": "numbers"
                        },
                        "var_type_annotated": "ARRAY"
                      }
,
                      "children": [
                        {

                          "node_type": "NUMBER",
                          "token": {
                              "type": "INTEGER_CONST",
                              "value": "1"
                          },
                          "var_type_annotated": "INT64",
                          "i_val": 0
                        }
                      ]
                    }
                  ]
                }

              },
              {

                "node_type": "EXPR_STMT",
                "var_type_annotated": "VOID",
                "left": 
                {

                  "node_type": "WRITELN",
                  "var_type_annotated": "UNKNOWN_VAR_TYPE",
                  "children": [
                    {

                      "node_type": "ARRAY_ACCESS",
                      "var_type_annotated": "STRING",
                      "left": 
                      {

                        "node_type": "VARIABLE",
                        "token": {
                            "type": "IDENTIFIER",
                            "value": "labels"
                        },
                        "var_type_annotated": "ARRAY"
                      }
,
                      "children": [
                        {

                          "node_type": "NUMBER",
                          "token": {
                              "type": "INTEGER_CONST",
                              "value": "0"
                          },
                          "var_type_annotated": "INT64",
                          "i_val": 0
                        }
                      ]
                    }
                  ]
                }

              },
              {

                "node_type": "RETURN",
                "token": {
                    "type": "RETURN",
                    "value": "return"
                },
                "var_type_annotated": "INT64",
                "left": 
                {

                  "node_type": "NUMBER",
                  "token": {
                      "type": "INTEGER_CONST",
                      "value": "0"
                  },
                  "var_type_annotated": "INT64",
                  "i_val": 0
                }

              }
            ]
          }

        }
      ]
    }
,
    "body": 
    {

      "node_type": "COMPOUND",
      "var_type_annotated": "VOID",
      "children": [
        {

          "node_type": "EXPR_STMT",
          "token": {
              "type": "IDENTIFIER",
              "value": "main"
          },
          "var_type_annotated": "VOID",
          "left": 
          {

            "node_type": "PROCEDURE_CALL",
            "token": {
                "type": "IDENTIFIER",
                "value": "main"
            },
            "var_type_annotated": "INT64"
          }

        }
      ]
    }

  }

}

